source/request-response/protocol_adapter.c

```diff
@@ -0,0 +1,428 @@
+/**
+ * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0.
+ */
+
+#include <aws/mqtt/private/request-response/protocol_adapter.h>
+
+#include <aws/io/event_loop.h>
+#include <aws/mqtt/private/client_impl_shared.h>
+#include <aws/mqtt/private/request-response/weak_ref.h>
+#include <aws/mqtt/private/v5/mqtt5_client_impl.h>
+#include <aws/mqtt/v5/mqtt5_client.h>
+#include <aws/mqtt/v5/mqtt5_listener.h>
+
+/*
+ * New API contract
+ *
+ * Invariant 1: Subscribe is only called from the RR subscription manager when going from 0 to 1 pending operations
+ * Invariant 2: Unsubscribe is only called from the RR subscription manager when there are 0 pending operations, not
+ *   necessarily on the exact transition to zero though.
+ *
+ * Entries are not tracked with the exception of eventstream impl which needs the stream handles to close.
+ * A subscribe failure should not trigger an unsubscribe, only notify the status callback.
```

Comment:
Should we include in here that we also don't retry a subscription? "A subscribe failure should not trigger an unsubscribe or retry a subscription, only notify the status callback." ?

Reply:
This is just me talking to myself.  If it creates confusion I'll just remove.

Reply:
Let's keep it in if we can. This sort of description seems extremely useful for anyone coming to this fresh.